### PR TITLE
[CSL-1249] Items API V2

### PR DESCRIPTION
**Pull request by esezen**

### CODE REVIEW

File: constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java [hunk 1]
```diff
@@ -205,114 +205,24 @@ public boolean verify() throws ConstructorException {
...
-            getResponseBody(response);
-            return true;
-        } catch (Exception exception) {
-            throw new ConstructorException(exception);
-        }
-    }
-
-    /**
-     * Adds multiple items to your autocomplete (limit of 1000 items)
+     * Adds multiple items to your index whilst updating existing ones (limit of 1000 items)
```
Comment (sblaurock) [inline constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java:208]: Add comma (`1,000`)

File: constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java [hunk 2]
```diff
@@ -205,114 +205,24 @@ public boolean verify() throws ConstructorException {
...
      *
      * @param items the items you want to add.
      * @param autocompleteSection the section of the autocomplete that you're adding the items to.
      * @return true if working
      * @throws ConstructorException if the request is invalid.
      */
-    public boolean addItemBatch(ConstructorItem[] items, String autocompleteSection) throws ConstructorException {
+    public boolean addOrUpdateItems(ConstructorItem[] items, String autocompleteSection) throws ConstructorException {
```
Comment (sblaurock) [inline constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java:215]: Can we change `autocompleteSection` to `section` or will that lead to inconsistencies across the repo?

File: constructorio-client/src/main/java/io/constructor/client/ConstructorVariation.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    private String itemId;
+    private String description;
+    private Map<String, String> facets;
+    private Map<String, String> metadata;
+    private List<String> groupIds;
+
+    /**
+     * Creates an autocomplete item.  Optional public fields are in the <a href="https://docs.constructor.io/rest-api.html#add-an-item">API documentation</a>
+     *
+     * @param name the name of the item that you are adding.
```
Comment (sblaurock) [inline constructorio-client/src/main/java/io/constructor/client/ConstructorVariation.java:27]: Should be `variation`

File: constructorio-client/src/main/java/io/constructor/client/ConstructorVariation.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        this.url = null;
+        this.imageUrl = null;
+        this.description = null;
+        this.facets = null;
+        this.metadata = null;
+        this.groupIds = null;
+    }
+
+    /**
+     * Returns the HashMap form of an autocomplete item for converting to JSON
```
Comment (sblaurock) [inline constructorio-client/src/main/java/io/constructor/client/ConstructorVariation.java:56]: Avoid "autocomplete item"

**Review by sblaurock — COMMENTED** [review]
Looks great - few small comments

**Comment by sblaurock** [comment]
Confirmed we have no active consumers of these methods at the moment - will be released as a minor change.

**Review by stanlp1 — APPROVED** [review]
LGTM!

File: constructorio-client/src/test/java/io/constructor/client/ConstructorVariationTest.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    public String getRandomProductName() {
+        return "Product" + UUID.randomUUID().toString().replaceAll("[\\s\\-()]", "");
+    }
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    @Test
+    public void newWithNullvariationNameShouldFail() throws Exception {
```
Comment (jjl014) [inline constructorio-client/src/test/java/io/constructor/client/ConstructorVariationTest.java:22]: Casing: `newWithNullVariationNameShouldFail`

File: constructorio-client/src/test/java/io/constructor/client/ConstructorItemTest.java [hunk 6]
```diff
@@ -19,29 +19,31 @@ public String getRandomProductName() {
     public ExpectedException thrown = ExpectedException.none();
 
     @Test
-    public void newWithNullItemNameShouldFail() throws Exception {
+    public void newWithNullItemNameShouldPass() throws Exception {
```
Comment (jjl014) [inline constructorio-client/src/test/java/io/constructor/client/ConstructorItemTest.java:22]: Should this be `newWithNullItemIdShouldPass`?

File: constructorio-client/src/test/java/io/constructor/client/ConstructorItemTest.java [hunk 7]
```diff
@@ -19,29 +19,31 @@ public String getRandomProductName() {
...
+        assertEquals(item.getName(), itemName);
         assertEquals(item.getSuggestedScore(), null);
         assertEquals(item.getKeywords(), null);
         assertEquals(item.getUrl(), null);
         assertEquals(item.getImageUrl(), null);
         assertEquals(item.getDescription(), null);
-        assertEquals(item.getId(), null);
+        assertEquals(item.getId(), itemName);
         assertEquals(item.getFacets(), null);
         assertEquals(item.getFacets(), null);
```
Comment (jjl014) [inline constructorio-client/src/test/java/io/constructor/client/ConstructorItemTest.java:48]: Let's remove this duplicate `getFacets()` call

File: constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java [hunk 8]
```diff
@@ -390,22 +367,37 @@ public boolean removeItemBatch(ConstructorItem[] items, String autocompleteSecti
...
+        return modifyVariations(variations, "Products", false);
+    }
+
+    public boolean modifyVariations(ConstructorVariation[] variations, String section) throws ConstructorException {
+        return modifyVariations(variations, section, false);
+    }
+
     /**
-     * Modifies an item from your autocomplete.
+     * Adds multiple variation to your index whilst updating existing ones (limit of 1,000 items)
```
Comment (jjl014) [inline constructorio-client/src/main/java/io/constructor/client/ConstructorIO.java:379]: `variation` => `variations`

**Review by jjl014 — APPROVED** [review]
Nice, this lgtm. Left a few comments but great job!

File: constructorio-client/src/test/java/io/constructor/client/ConstructorItemTest.java (hunk 6 above)
Comment (esezen) [reply constructorio-client/src/test/java/io/constructor/client/ConstructorItemTest.java:22]: It should have been `newWithNullItemIdShouldFail`.
Great catch!

**Pull request merged**